### PR TITLE
Merge expression types and native expression types

### PR DESCRIPTION
**Pull request by rajyan**
resolving todos in https://github.com/phpstan/phpstan/issues/8191

### CODE REVIEW

**Comment by rajyan** [comment]
After resolving the inconsistencies, I think it might be better to have `nativeType` and `nativeCertainty` in `ExpressionTypeHolder` and merge nativeExpression types. With it, we can merge some duplicate of  code for native types which can improve performance in some cases.

**Comment by rajyan** [comment]
I realized that, the current conditionalExpressionTypes logic can make the count of native and expression types differ.

File: src/Analyser/MutatingScope.php [hunk 1]
```diff
@@ -3861,8 +3963,10 @@ public function filterBySpecifiedTypes(SpecifiedTypes $specifiedTypes): self
 
 			if ($typeHolder->getCertainty()->no()) {
 				unset($scope->expressionTypes[$variableExprString]);
+				unset($scope->nativeExpressionTypes[$variableExprString]);
 			} else {
 				$scope->expressionTypes[$variableExprString] = $typeHolder;
+				$scope->nativeExpressionTypes[$variableExprString] = $typeHolder;
```
Comment (rajyan) [inline src/Analyser/MutatingScope.php:3966]: > I realized that, the current conditionalExpressionTypes logic can make the count of native and expression types differ.

this is the incorrect one. Added just to see the temporal validation passes.

**Comment by rajyan** [comment]
On second thought I think https://github.com/phpstan/phpstan-src/pull/1951#issuecomment-1299371448 is possible. I would clean up this PR first.

**Comment by rajyan** [comment]
If expressionType exists for some expression, nativeExpressionType should always exist too. Although the certainty can be different.

**Comment by ondrejmirtes** [comment]
> After resolving the inconsistencies, I think it might be better to have nativeType and nativeCertainty in ExpressionTypeHolder

My problem is that currently it's not possible to have an entry in these arrays with certainty=no.

File: src/Analyser/MutatingScope.php [hunk 2]
```diff
@@ -365,6 +372,40 @@ public function afterClearstatcacheCall(): self
...
 			}
 		}
+		$nativeExpressionTypes = $this->nativeExpressionTypes;
+		foreach (array_keys($nativeExpressionTypes) as $exprString) {
+			// list from https://www.php.net/manual/en/function.clearstatcache.php
+
+			// stat(), lstat(), file_exists(), is_writable(), is_readable(), is_executable(), is_file(), is_dir(), is_link(), filectime(), fileatime(), filemtime(), fileinode(), filegroup(), fileowner(), filesize(), filetype(), and fileperms().
+			foreach ([
+						 'stat',
+						 'lstat',
```
Comment (ondrejmirtes) [inline src/Analyser/MutatingScope.php:382]: This is exactly the kind of duplication I'd like to avoid :)

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Analyser/MutatingScope.php (hunk 2 above)
Comment (rajyan) [reply src/Analyser/MutatingScope.php:382]: Definitely. I just wanted to test that it is possible to make count of native and full type same, if we add them everywhere.

**Comment by rajyan** [comment]
> My problem is that currently it's not possible to have an entry in these arrays with certainty=no.

Yeah, I think we currently don't have a situation that expressionTypes holds certainty "no", because we are unsetting the expression when invalidated.

File: src/Analyser/MutatingScope.php [hunk 3]
```diff
@@ -248,6 +248,13 @@ public function enterDeclareStrictTypes(): self
 			null,
 			null,
 			$this->expressionTypes,
+			[],
+			null,
+			null,
+			true,
+			[],
+			[],
+			$this->nativeExpressionTypes,
```
Comment (rajyan) [inline src/Analyser/MutatingScope.php:257]: maybe this one is not correct too. There should be no situation that native type assigned before declare

**Comment by rajyan** [comment]
Sorry, but I would close this one for now.

**Pull request closed without merging**